solutions/USACO dec19-milkvisits-silver.mdx

```diff
@@ -0,0 +1,224 @@
+---
+id: usaco-968
+source: USACO Silver 2019 December
+title: Milk Visits
+author: Qi Wang
+---
+
+[Official Analysis (Offline)](http://www.usaco.org/current/data/sol_milkvisits_silver_dec19.html)
+
+**Time Complexity:** $\mathcal{O}(N + M)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+//Official Solution
+#include <bits/stdc++.h>
+using namespace std;
+
+typedef vector<int> vi; 
+ 
+#define FOR(i,a,b) for (int i = (a); i < (b); ++i)
+#define F0R(i,a) FOR(i,0,a)
+#define ROF(i,a,b) for (int i = (b)-1; i >= (a); --i)
+#define R0F(i,a) ROF(i,0,a)
+#define trav(a,x) for (auto& a: x)
+ 
+#define pb push_back
+ 
+void setIO(string name) {
+	ios_base::sync_with_stdio(0); cin.tie(0);
+	freopen((name+".in").c_str(),"r",stdin);
+	freopen((name+".out").c_str(),"w",stdout);
+}
+
+const int MX = 100005;
+
+int N,M,comp[MX],num;
+char col[MX];
+vi adj[MX];
+ 
+void dfs(int x) {
+	if (comp[x]) return;
+	comp[x] = num;
+	trav(t,adj[x]) if (col[t] == col[x]) dfs(t);
+}
+ 
+int main() {
+	setIO("milkvisits"); cin >> N >> M;
+	string s; cin >> s;
+	FOR(i,1,N+1) col[i] = s[i-1];
+	F0R(i,N-1) {
+		int A,B; cin >> A >> B;
+		adj[A].pb(B), adj[B].pb(A);
+	}
+	FOR(i,1,N+1) if (!comp[i]) {
+		num ++;
+		dfs(i);
+	}
+	F0R(i,M) {
+		int A,B; char C; cin >> A >> B >> C;
+		if (col[A] == C || comp[A] != comp[B]) cout << 1;
+		else cout << 0;
+	}
+	cout << "\n";
+}
+```
+
+</CPPSection>
+
+<JavaSection>
```

Comment:
cpp version is so much cleaner (such as avoiding vis array and single if condition in the answer), can we replicate it? (btw why are they so different)

Reply:
mbmb the cpp sol is the USACO sol, which explains the diff. sry for a redundant java sol tho, didn't compare with the usaco tutorial last night!

Reply:
Oh, no worries! Just thought that we might as well make them more similar so people wouldn't feel the need to switch to different langs to see different ways to implement it, as well as the fact that both solutions can use improvements from the other. Also if variable names are consistent, two people looking at different langs can talk about the problem/solution in a way that makes sense.

Reply:
K cool, ill start editing the sol